Server panic when triggering model-backend fails in sync mode

```
{"level":"error","ts":1666169847.2963574,"caller":"service/service.go:391","msg":"[model-backend] Error TriggerModel at 0th model instance models/yolov4/instances/v1.0-gpu: rpc error: code = FailedPrecondition desc = [handler] inference model error","stacktrace":"github.com/instill-ai/pipeline-backend/pkg/service.(*service).TriggerPipeline.func1\n\t/src/pkg/service/service.go:391"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0xf3750d]

goroutine 88 [running]:
github.com/instill-ai/pipeline-backend/pkg/service.(*service).TriggerPipeline.func1()
	/src/pkg/service/service.go:394 +0x32d
created by github.com/instill-ai/pipeline-backend/pkg/service.(*service).TriggerPipeline
	/src/pkg/service/service.go:381 +0x88c
```

--- a/pkg/service/service.go
+++ b/pkg/service/service.go
@@ -378,7 +378,10 @@ func (s *service) TriggerPipeline(req *pipelinePB.TriggerPipelineRequest, dbPipe
 	wg.Add(1)
 
 	var modelInstOutputs []*pipelinePB.ModelInstanceOutput
+	errors := make(chan error)
 	go func() {
+		defer wg.Done()
+
 		for idx, modelInstance := range dbPipeline.Recipe.ModelInstances {
 
 			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
@@ -388,7 +391,9 @@ func (s *service) TriggerPipeline(req *pipelinePB.TriggerPipelineRequest, dbPipe
 				Inputs: inputs,
 			})
 			if err != nil {
+				errors <- err
 				logger.Error(fmt.Sprintf("[model-backend] Error %s at %dth model instance %s: %v", "TriggerModel", idx, modelInstance, err.Error()))
+				return
 			}
 
 			taskOutputs := cvtModelTaskOutputToPipelineTaskOutput(resp.TaskOutputs)
@@ -405,7 +410,9 @@ func (s *service) TriggerPipeline(req *pipelinePB.TriggerPipelineRequest, dbPipe
 			// Increment trigger image numbers
 			uid, err := resource.GetPermalinkUID(dbPipeline.Owner)
 			if err != nil {
+				errors <- err
 				logger.Error(err.Error())
+				return
 			}
 			if strings.HasPrefix(dbPipeline.Owner, "users/") {
 				s.redisClient.IncrBy(context.Background(), fmt.Sprintf("user:%s:trigger.image.num", uid), int64(len(inputs)))
@@ -413,13 +420,23 @@ func (s *service) TriggerPipeline(req *pipelinePB.TriggerPipelineRequest, dbPipe
 				s.redisClient.IncrBy(context.Background(), fmt.Sprintf("org:%s:trigger.image.num", uid), int64(len(inputs)))
 			}
 		}
-		wg.Done()
+		errors <- nil
 	}()
+
+	go func() {
+		wg.Wait()
+		close(errors)
+	}()
+
+	for err := range errors {
+		if err != nil {
+			return nil, fmt.Errorf("error when trigger model instance %v", err.Error())
+		}
+	}
 
 	switch {
 	// If this is a SYNC trigger (i.e., HTTP, gRPC source and destination connectors), return right away
 	case dbPipeline.Mode == datamodel.PipelineMode(pipelinePB.Pipeline_MODE_SYNC):
-		wg.Wait()
 		return &pipelinePB.TriggerPipelineResponse{
 			DataMappingIndices:   dataMappingIndices,
 			ModelInstanceOutputs: modelInstOutputs,
@@ -427,7 +444,6 @@ func (s *service) TriggerPipeline(req *pipelinePB.TriggerPipelineRequest, dbPipe
 	// If this is a async trigger, write to the destination connector
 	case dbPipeline.Mode == datamodel.PipelineMode(pipelinePB.Pipeline_MODE_ASYNC):
 		go func() {
-			wg.Wait()
 			for idx, modelInstRecName := range dbPipeline.Recipe.ModelInstances {
 				ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
 				defer cancel()

--- a/pkg/service/service.go
+++ b/pkg/service/service.go
@@ -454,6 +454,7 @@ func (s *service) TriggerPipeline(req *pipelinePB.TriggerPipelineRequest, dbPipe
 			for err := range errors {
 				if err != nil {
 					logger.Error(fmt.Sprintf("[connector-backend] Error trigger model instance got error %v", err.Error()))
+					return
 				}
 			}
 

--- a/pkg/service/service.go
+++ b/pkg/service/service.go
@@ -453,7 +453,7 @@ func (s *service) TriggerPipeline(req *pipelinePB.TriggerPipelineRequest, dbPipe
 			}()
 			for err := range errors {
 				if err != nil {
-					logger.Error(fmt.Sprintf("[connector-backend] Error trigger model instance got error %v", err.Error()))
+					logger.Error(fmt.Sprintf("[model-backend] Error trigger model instance got error %v", err.Error()))
 					return
 				}
 			}

--- a/pkg/service/service.go
+++ b/pkg/service/service.go
@@ -384,7 +384,7 @@ func (s *service) TriggerPipeline(req *pipelinePB.TriggerPipelineRequest, dbPipe
 
 		for idx, modelInstance := range dbPipeline.Recipe.ModelInstances {
 
-			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+			ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
 			defer cancel()
 			resp, err := s.modelServiceClient.TriggerModelInstance(ctx, &modelPB.TriggerModelInstanceRequest{
 				Name:   modelInstance,

--- a/pkg/service/service.go
+++ b/pkg/service/service.go
@@ -430,7 +430,12 @@ func (s *service) TriggerPipeline(req *pipelinePB.TriggerPipelineRequest, dbPipe
 
 	for err := range errors {
 		if err != nil {
-			return nil, fmt.Errorf("error when trigger model instance %v", err.Error())
+			switch {
+			case strings.Contains(err.Error(), "code = DeadlineExceeded"):
+				return nil, status.Errorf(codes.DeadlineExceeded, "trigger model instance got timeout error")
+			default:
+				return nil, status.Errorf(codes.Internal, fmt.Sprintf("trigger model instance got error %v", err.Error()))
+			}
 		}
 	}
 

--- a/pkg/service/service.go
+++ b/pkg/service/service.go
@@ -384,7 +384,7 @@ func (s *service) TriggerPipeline(req *pipelinePB.TriggerPipelineRequest, dbPipe
 
 		for idx, modelInstance := range dbPipeline.Recipe.ModelInstances {
 
-			ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
 			defer cancel()
 			resp, err := s.modelServiceClient.TriggerModelInstance(ctx, &modelPB.TriggerModelInstanceRequest{
 				Name:   modelInstance,
@@ -423,25 +423,23 @@ func (s *service) TriggerPipeline(req *pipelinePB.TriggerPipelineRequest, dbPipe
 		errors <- nil
 	}()
 
-	go func() {
-		wg.Wait()
-		close(errors)
-	}()
-
-	for err := range errors {
-		if err != nil {
-			switch {
-			case strings.Contains(err.Error(), "code = DeadlineExceeded"):
-				return nil, status.Errorf(codes.DeadlineExceeded, "trigger model instance got timeout error")
-			default:
-				return nil, status.Errorf(codes.Internal, fmt.Sprintf("trigger model instance got error %v", err.Error()))
-			}
-		}
-	}
-
 	switch {
 	// If this is a SYNC trigger (i.e., HTTP, gRPC source and destination connectors), return right away
 	case dbPipeline.Mode == datamodel.PipelineMode(pipelinePB.Pipeline_MODE_SYNC):
+		go func() {
+			wg.Wait()
+			close(errors)
+		}()
+		for err := range errors {
+			if err != nil {
+				switch {
+				case strings.Contains(err.Error(), "code = DeadlineExceeded"):
+					return nil, status.Errorf(codes.DeadlineExceeded, "trigger model instance got timeout error")
+				default:
+					return nil, status.Errorf(codes.Internal, fmt.Sprintf("trigger model instance got error %v", err.Error()))
+				}
+			}
+		}
 		return &pipelinePB.TriggerPipelineResponse{
 			DataMappingIndices:   dataMappingIndices,
 			ModelInstanceOutputs: modelInstOutputs,
@@ -449,6 +447,16 @@ func (s *service) TriggerPipeline(req *pipelinePB.TriggerPipelineRequest, dbPipe
 	// If this is a async trigger, write to the destination connector
 	case dbPipeline.Mode == datamodel.PipelineMode(pipelinePB.Pipeline_MODE_ASYNC):
 		go func() {
+			go func() {
+				wg.Wait()
+				close(errors)
+			}()
+			for err := range errors {
+				if err != nil {
+					logger.Error(fmt.Sprintf("[connector-backend] Error trigger model instance got error %v", err.Error()))
+				}
+			}
+
 			for idx, modelInstRecName := range dbPipeline.Recipe.ModelInstances {
 				ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
 				defer cancel()

